lib/linked_list.rb

```diff
@@ -117,33 +201,57 @@ def has_cycle
     # Additional Exercises 
     # returns the value in the first node
     # returns nil if the list is empty
-    # Time Complexity:  
-    # Space Complexity
+    # Time Complexity: O(1)
+    # Space Complexity O(1)
     def get_first
-      raise NotImplementedError
+      return nil if @head == nil
+      return @head.data
     end
 
     # method that inserts a given value as a new last node in the linked list
-    # Time Complexity:  
-    # Space Complexity
+    # Time Complexity:  O(n), where n is the length of the list
+    # Space Complexity: O(1)
     def add_last(value)
-      raise NotImplementedError
+      if @head == nil
+        @head = Node.new(value, nil)
+      else
+        current = @head
+        while current.next != nil
+          current = current.next
+        end
+
+        current.next = Node.new(value, nil)
+      end
     end
 
     # method that returns the value of the last node in the linked list
     # returns nil if the linked list is empty
-    # Time Complexity:  
-    # Space Complexity
+    # Time Complexity:  O(n), where n is the length of the list
+    # Space Complexity: O(1)
     def get_last
-      raise NotImplementedError
+      return nil if @head == nil
+      current = @head
+        while current.next != nil
+          current = current.next
+        end
+      
+      return current.data
     end
 
     # method to insert a new node with specific data value, assuming the linked
     # list is sorted in ascending order
-    # Time Complexity:  
-    # Space Complexity
+    # Time Complexity:  O(n), where n is the length of the list
+    # Space Complexity: O(1)
     def insert_ascending(value)
-      raise NotImplementedError
+      add_first(value) if @head == nil

```

Comment:
This doesn't look right, you would end up inserting value into the list multiple times.